apps/pipeline-visualization/src/mermaid_builder.ts

```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { BlockDefinition } from '@jvalue/jayvee-language-server';
+
+export class MermaidBuilder {
+  protected mermaidLines: string[] = [];
+
+  pipelineHead(pipelineName: string): MermaidBuilder {
+    this.mermaidLines.push('subgraph ' + pipelineName);
+    return this;
+  }
+
+  blockType(block: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i>]`,
+    );
+    return this;
+  }
+
+  blockTypeWithProperties(
+    block: BlockDefinition,
+    propertyString: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${block.name}[${block.name}<br><i>${block.type.ref?.name}</i><br>${propertyString}]`,
+    );
+    return this;
+  }
+
+  branching(block: BlockDefinition, child: BlockDefinition): MermaidBuilder {
+    this.mermaidLines.push(`${block.name}-->${child.name}`);
+    return this;
+  }
+
+  chain(chain: string[]): MermaidBuilder {
+    this.mermaidLines.push(chain.join('-->'));
+    return this;
+  }
+
+  compositeHeader(
+    block: BlockDefinition,
+    depth: number,
+    subgraphDirection: string,
+  ): MermaidBuilder {
+    this.mermaidLines.push(
+      `${'\t'.repeat(depth)}subgraph ${block.name}\n ${'\t'.repeat(
+        depth,
+      )}direction ${subgraphDirection}`,
+    );
+    return this;
+  }
+
+  compositeBody(subblocks: string[], depth: number): MermaidBuilder {
+    this.mermaidLines.push(`${'\t'.repeat(depth)}` + subblocks.join('-->'));
+    return this;
+  }
+
+  compositeTail(depth: number): MermaidBuilder {
+    this.mermaidLines.push(`${'\t'.repeat(depth)}end`);
+    return this;
+  }
+
+  classAssign(block: BlockDefinition, className: string): MermaidBuilder {
+    this.mermaidLines.push(`class ${block.name} ${className};`);
+    return this;
+  }
+
+  section(section: string): MermaidBuilder {
+    this.mermaidLines.push(section);
+    return this;
+  }
+
+  end(): MermaidBuilder {
+    this.mermaidLines.push('end');
+    return this;
+  }
+
+  newLine(): MermaidBuilder {
```

Comment:
This is not the job of a caller of a mermaid builder.

When deciding if a method should be available to a caller, try asking yourself this question:
> Does it make sense to _do X_ when building a diagram?

E.g. "Does it make sense to insert a new line when building a diagram?" - Not really, right?
However, it does make sense for a mermaid diagram *Builder* to produce a pretty output with a new line. Therefore, this method is useful for internal use only.

Reply:
ok, I could make it a protected method
But how do I achieve the same behavior? Currently, the caller knows in which cases I want a new line after a section.

Reply:
split into different parts of members and connect them with new-line